### ISSUE TITLE
Empty tuple type

### ISSUE THREAD

**Issue by nickie** [Microsoft/TypeScript#7520]
According to the specifications ([3.8.5](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#385-tuple-type-literals)), the following two should be illegal.  Tuple types should not be empty and there should be no trailing comma.  However, the implementation seems to allow them.

    var t1 : [];
    var t2 : [number,];

If they are to be allowed and the specifications are to be revised, let me suggest that the tuple type be generalized in the same way as array literals are, in JS.  E.g.

    var t: [, number,, string,,];

where `t` is a 5-tuple whose 2nd element is a `number` and whose 4th element is `string` (the other three are unknown --- `any`).

**Comment by RyanCavanaugh** [comment]
This is what I get when I compile this:

> (1,10): error TS1122: A tuple type element list cannot be empty.
> (2,17): error TS1009: Trailing comma not allowed.

What version are you running?

**Comment by nickie** [comment]
Sorry, I should have mentioned this.  Version 1.9.0, from a (slightly outdated, 6 days ago) git clone, commit f06423b.  It seems that the current version (commit 9f6f871) issues the errors that you mentioned, so it seems to have been fixed in the last 6 days.  Sorry about the false alarm.

On the other hand, I really wouldn't mind allowing these two and adopting the more liberal view of tuple types that I suggested above.  I'm leaving this open, in case somebody wants to think about it...

**Comment by nickie** [comment]
Oops, sorry, I was wrong once again.  The problem seems to be still there in the current commit 9f6f871.

**Comment by RyanCavanaugh** [comment]
I think something's wrong with whatever setup you're using to test this. I'm on the same commit and am seeing the errors.

If you want to open a separate suggestion for gapped tuple types, please do, but I'd rather not have this issue tracking it since it's logged as if it were a bug.

**Comment by nickie** [comment]
The issue seems to be more complicated than I thought.  My setup seems to be OK (`jake runtests` completes successfully).

In fact, I was testing a slightly larger example containing these two lines.  With that larger example, there were no errors reported for these two lines, but there were errors in lines following these two.  Here's a MWE showing this strange behaviour:

    var t1 : [];
    var t3 : [number,,];

Now (in the current commit 1156141), although both lines are erroneous, only the second one produces an error message:

    binding-patterns.ts(2,18): error TS1110: Type expected.

**Comment by RyanCavanaugh** [comment]
That's normal behavior. When you have a syntax error, we don't issue type errors. A missing entry in a delimited list is a syntax error, but an empty tuple type is a type error (as you can see this distinction is sometimes a bit arbitrary, but it usually makes sense).

You can see the same behavior e.g.:
```ts
var t1 : [];
var y = 1 & ; // Syntax error here, so no type error issued on line 1
```

**Comment by nickie** [comment]
OK, understood.  My mistake then...  Issue closed.
